cloudify_agent/operations.py

```diff
@@ -165,3 +182,98 @@ def _save_daemon(daemon):
         username=utils.internal.get_daemon_user(),
         storage=utils.internal.get_daemon_storage_dir())
     factory.save(daemon)
+
+
+def _get_broker_url(ctx, agent):
+    agent = copy.deepcopy(agent)
+    broker = _get_broker_config(ctx)
+    for k, v in broker.iteritems():
+        if not agent.get(k):
+            agent[k] = v
+    return utils.internal.get_broker_url(agent)
+
+
+def _generate_new_agent_name(old_agent_name, suffix):
```

Comment:
This is a nice function. I think we should replace the current internal.utils.generate_agent_name() with something similar to this. at the very least lets replace the call to  internal.utils.generate_agent_name() in the restart operation with this call.